commit 9af7ce8cc320dd7fca3639136e5b7011b27aa5f2
Author:     Yuhong Wen <yuhongw@nvidia.com>
AuthorDate: Wed Jun 26 10:13:39 2024 -0400
Commit:     GitHub <noreply@github.com>
CommitDate: Wed Jun 26 10:13:39 2024 -0400

    Allow the simulator to pass the End_Run aux message to clients. (#2653)
---
 nvflare/private/fed/server/server_engine.py | 35 ++++++++++++++++-------------
 1 file changed, 19 insertions(+), 16 deletions(-)

diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index 47fc3a5e..96663dd1 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -508,25 +508,28 @@ class ServerEngine(ServerEngineInternalSpec):
         secure=False,
     ) -> dict:
         try:
-            if not targets:
-                targets = []
-                for t in self.get_clients():
-                    targets.append(t.name)
-            if targets:
-                return self.run_manager.aux_runner.send_aux_request(
-                    targets=targets,
-                    topic=topic,
-                    request=request,
-                    timeout=timeout,
-                    fl_ctx=fl_ctx,
-                    optional=optional,
-                    secure=secure,
-                )
-            else:
-                return {}
+            return self.send_aux_to_targets(targets, topic, request, timeout, fl_ctx, optional, secure)
         except Exception as e:
             self.logger.error(f"Failed to send the aux_message: {topic} with exception: {secure_format_exception(e)}.")
 
+    def send_aux_to_targets(self, targets, topic, request, timeout, fl_ctx, optional, secure):
+        if not targets:
+            targets = []
+            for t in self.get_clients():
+                targets.append(t.name)
+        if targets:
+            return self.run_manager.aux_runner.send_aux_request(
+                targets=targets,
+                topic=topic,
+                request=request,
+                timeout=timeout,
+                fl_ctx=fl_ctx,
+                optional=optional,
+                secure=secure,
+            )
+        else:
+            return {}
+
     def sync_clients_from_main_process(self):
         # repeatedly ask the parent process to get participating clients until we receive the result
         # or timed out after 30 secs (already tried 30 times).
